### ISSUE TITLE
motor position

### ISSUE THREAD

**Issue by bertca** [ridiekel/jeletask#4]
Hi,

I am currently looking at the motor function. If I monitor the mqtt messages when I change the position of one of my covers, I get a state of 0 (up) or 100 (down). The stop command is just a repeat of the last issued command (again 0 or 100). In the most recent versions op the Teletask API documentation and the Teletask software, you can set the cover position to a certain percentage. If I understand the Teletask documentation well, the changed position of the cover is send by the Micros+ central. Is there way to capture this? It would be nice to get the actual cover position in Home Assistant.

Please let me know if this is feasible, if  my view is completely wrong, etc...

Thanks in advance

![Knipsel](https://user-images.githubusercontent.com/79922139/161374154-1cbb8255-413c-43a9-a25d-46dfc8f1e044.JPG)

**Comment by bertca** [comment]
anyone?

**Comment by ridiekel** [comment]
If I understand correctly, this is only about receiving the events? In the docs I cannot find anything on setting the position, correct?

**Comment by bertca** [comment]
I assume that something similar as dimmers is possible. I added the documentation I have from Teletask. It is mentioning a "motor_go_to_position" data parameter. But you have to enable the "position indication" in Prosoft for that.
Again, I'm not that familiar with all this. But it would really be nice to get HA and Teletask exchanging more data concerning the cover state/position/...
MQTT auto discovery of covers in Home Assistant would also be nice ;-)

[tds15132-library.pdf](https://github.com/ridiekel/jeletask/files/9294873/tds15132-library.pdf)
.

**Comment by ridiekel** [comment]
I missed that part in the docs of TT. I will look into it. Do you need a special hardware module for the motor positions? I have a motor module, but don't know if it supports positioning.

In either case, it should be possible to map this. It's just going to take me some time to get back into to TT codebase mindset. I will try to also write some tests, so I don't need the TT hardware. I will keep youo posted on the progress.

**Comment by bertca** [comment]
I have Teletask motor modules (the motor on the cover is a standard thing without specific intelligence). In Prosoft you have to go to the motor interface settings, select a motor and go to the advanced tab. There you see the "position setting" where you can select "indication for all positions" (see picture - sorry in Dutch). Once you have done that, you see a percentage indication in the Teletask Gui (other picture - Eetplaats motor).
If I can be of further assistance, please let me know. And take your time - this is not a matter of life and death ;-)

![tt_prosoft](https://user-images.githubusercontent.com/79922139/183831058-964a2ddf-398f-486e-8800-f67a2eca0e0f.JPG)
![tt_isgui](https://user-images.githubusercontent.com/79922139/183831083-8c59315d-e82c-4683-801f-4e9f91fb88f0.JPG)

**Comment by ridiekel** [comment]
Cool, I have the same module, so I can probably experiment with this. Did not know TT had support for this.

**Comment by SanderM2** [comment]
I'm not a programmer so no code should be trusted here but I created a quick&dirty 'hack' to get the motor positions:

Changes were made in /client/src/main/java/io/github/ridiekel/jeletask/client/builder/composer/v3_1/MicrosPlusCommandConfiguration.java

Original parse() function:

```
        public EventMessage parse(CentralUnit config, MessageHandler messageHandler, byte[] rawBytes, byte[] payload) {
            Function function = messageHandler.getFunction(payload[1]);
            int number = this.getOutputNumber(messageHandler, payload, 2);
            String state = getState(messageHandler, config, function, number, payload, 5);
            return new EventMessage(config, rawBytes, function, number, state);
        }
```

Modified parse() function:
```
        public EventMessage parse(CentralUnit config, MessageHandler messageHandler, byte[] rawBytes, byte[] payload) {
            Function function = messageHandler.getFunction(payload[1]);
            int number = this.getOutputNumber(messageHandler, payload, 2);

            int function_number = messageHandler.getFunctionConfig(function).getNumber();
            String state = ""; // wtf
            int start_index = 5; // original 
            if (function_number == 6) {
                // 6 = MOTOR
                // payload[5] = direction (current/last running direction)
                // payload[6] = power (-1= running && 0 = stopped?)
                // payload[7] = protection (see pdf)
                // payload[8] = position (current/last position). When motor is running it's the target position?
                // payload[9] = ... when motor is running the starting position is at this position?
                // payload[10-...] = What are those 5 extra bytes teletask adds? Maybe reverse engineering required. PDF lacks info 
                start_index=8;
                if (payload.length >= 10) {
                    // motor is running. Current position is at index 9 now
                    start_index = 9;
                }
                state = getState(messageHandler, config, function, number, payload, start_index);
            } else {
                state = getState(messageHandler, config, function, number, payload, start_index);
            }
            return new EventMessage(config, rawBytes, function, number, state);
        }
```

**Comment by ridiekel** [comment]
Just a quick update. I have been working on this, but the "API" of TT has changed a little bit since I started this project more than 10 years ago. Seems like they support some more advanced use cases, but this requires some rework in the core of the library I wrote. 
I know the things I need to do, but it will take more time than initially expected. 
One of the things that makes this more complicated, is that I also support the old Micros (with the network addon, instead of the plus). I'm a bit in dubio about still supporting it in the library, since it is also no longer supported by TT (the last version of their own software that works with it is 2.8). This means that the API's are diverging quite a bit, which means supporting both in the same code base is quite challenging.

I will keep you posted once I have completed a first version with the new API.

**Comment by SanderM2** [comment]
Do you still have the old micros yourself? Or do you have another reason for still supporting it? I think that at some point giving up support for old devices is needed. Especially if the manufacturer is not even supporting it anymore. (They have dropped support for a reason too, right...). If you still have the old device, I certainly understand why you want to support it.

If not, then I would seriously consider dropping support for it if that makes things less complicated. You could split the code, leaving the current state for people with the old micros and start a renewed project without support for the old micros. The new library can then move on with time and support the more advanced use cases.

**Comment by ridiekel** [comment]
I will check in the final solution if I need to drop support for it. I will then make a major version bump so it's clear you need to use the old containers for accessing the old micros HW.

**Comment by SanderM2** [comment]
FYI:

The current teletask app shows a percentage while the motor is spinning. I thought that those 5 undocumented additional bytes might be used for timing it. So, I asked Teletask about them and I got the following response:

"Voor de motor zijn er inderdaad enkele ongedocumenteerde 'bytes'.
Indien er 5 extra bytes zijn, wil dit zeggen dat de aan het lopen is én dat er info kan gegeven worden wanneer de positie (Position = target) bereikt zal zijn.
De 5 bytes zijn als volgt:
1: De huidige positie van de motor
2+3: Aantal seconden (*100) die de motor er nog zal overdoen om de positie te bereiken.
 4: eventuele correctie bij 0% in seconden
5: eventuele correctie bij 1000% in seconden"



I use this feature in the teletask app to open/close my covers partly when I'm not at home. So it would be nice if we can somehow use this data and make HA behave the same way.

But this is probably a new feature request?

**Comment by ridiekel** [comment]
Hi all,

I just published a new version 4.0.1 in dockerhub. You should now be able to send a motor to a position using mqtt. If anyone can test this, that would be great. You can find more info in the branch readme: https://github.com/ridiekel/jeletask/tree/4-motor-position

I have not yet pushed this to master, waiting first for confirmation. Also, I have not done the integration with HA for now. That is next on my agenda. But at least the mqtt integration should already work.

I'll keep you posted on any other updates.

**Comment by SanderM2** [comment]
I have tested 4.0.1 and it all seems to work fine.

Tested sending motor positions: Works
Tested ON/OFF: Works
Checked if new position was correctly seen when motor posion was changed externally: Works.

**Comment by SanderM2** [comment]
You said HA integration is next on your agenda.
Are you going to implement auto discovery for motors?

**Comment by ridiekel** [comment]
Yes, the plan is that auto discovery would work for motors also

**Comment by claesmathias** [comment]
I see the new release provides a lot of logs ;)

![Screenshot 2022-09-09 at 18 07 14](https://user-images.githubusercontent.com/1142852/189393927-4828ffe3-e97c-47b0-8079-5f07c8b8c02f.png)

**Comment by claesmathias** [comment]
HA auto-discovery is also bit messed up :(
![Screenshot 2022-09-09 at 18 09 44](https://user-images.githubusercontent.com/1142852/189394352-2ed264bb-db01-4e9f-ab21-8bf7661124f1.png)

**Comment by SanderM2** [comment]
In HA this works:

mqtt:
    cover:
        - name: "Gordijn 1"
          position_topic: "teletask2mqtt/teletask/motor/3/state"
          set_position_topic: "teletask2mqtt/teletask/motor/3/set"
          qos: 0
          position_open: 0
          position_closed: 100
          optimistic: false
          position_template: "{{ value_json.position }}"


It's pretty basic. It only shows position of the cover when you click on it. 
I guess that's normal in HA? And we need an integration if we want to see the position from a dashboard view.

**Comment by SanderM2** [comment]
I also have horizontal blinds in my home. These are motorized too.

In teletask there's an option "set slat position" that will activate the motor for just a second or so. It's configurable under the motor interface and the "set slat position" is only available from within a general/local mood.

I asked Teletask if the API supports setting the slat position too and they wrote back:

"Voor de positie gebruik je normaal voor de achtste byte (action) ‘11’, als je deze op ‘12’ zet stuur je de lamellen aan. De volgende byte is dan de lamelpositie."

Maybe this can be added too?

**Comment by SanderM2** [comment]
Now that we have "seconds_to_finish" (which actually is centiseconds) we could publish MQTT state topics to actually update the slider in HA in realtime like with the original teletask app.

Is this something you plan on adding? if not I'm going to write a separate script that catches the MQTT topic with seconds_to_finish and published new MQTT topics to get the slider in HA moving...

**Comment by bertca** [comment]
Hi Ridiekel,
I'm glad there is some progress. Unfortunately I'm not able to test right now due to lack of time. I will also do some tests as soon as I can.
Thank for the exellent job you're doing.

**Comment by claesmathias** [comment]
Hi,
I tried to re-install docker container [latest-arm64](https://hub.docker.com/layers/ridiekel/jeletask2mqtt/latest-arm64/images/sha256-8b2454953432748e6b08b3e4418d14b47fcd11aa389ebdfeaea8a32c32e5a3f1?context=explore), I see output is now in json format. Trying to control dimming lights from HA isn't working well. So rolled-back version 3.3.6. Let me know if you have the same behaviour.

**Comment by SanderM2** [comment]
Lights and switches are broken in 4.0.1 but a fix is in the 4-motor-position branch if you want to compile it yourself.
But I'm sure ridiekel will release a new version on dockerhub soon.

**Comment by claesmathias** [comment]
I pulled the new container image and still experience regression. Dimmed lights are auto discovered in HA, normal lights not. I will start some debugging in the weekend.

**Comment by SanderM2** [comment]
Are you using TELETASK_MQTT_RETAINED ? If so, what happens if you manually delete the matt topics and restart jeletask?

**Comment by claesmathias** [comment]
Yes, non-dimmable light are indeed relays. I don't use the flag TELETASK_MQTT_RETAINED, once I configured the experience was better in HA, I will test more in the weekend. Next to HA I use homebridge, which seems reconfiguration to catch the state of dimmers and temperature sensor.

![Screenshot 2022-09-21 at 22 31 31](https://user-images.githubusercontent.com/1142852/191604528-54abebe4-a974-4c3f-a129-a0fed5d2910f.png)


`        {
            "type": "lightbulb-OnOff",
            "name": "Voortuin",
            "topics": {
                "getOn": "teletask2mqtt/my_teletask/relay/18/state",
                "setOn": "teletask2mqtt/my_teletask/relay/18/set"
            },
            "onValue": "ON",
            "offValue": "OFF",
            "accessory": "mqttthing"
        },
        {
            "type": "lightbulb",
            "name": "Nachthal",
            "topics": {
                "setBrightness": "teletask2mqtt/my_teletask/dimmer/3/set",
                "getBrightness": "teletask2mqtt/my_teletask/dimmer/3/state"
            },
            "accessory": "mqttthing"
        },
        {
            "type": "lightbulb",
            "name": "Slaapkamer 1",
            "topics": {
                "setBrightness": "teletask2mqtt/my_teletask/dimmer/4/set",
                "getBrightness": "teletask2mqtt/my_teletask/dimmer/4/state"
            },
            "accessory": "mqttthing"
        },
        {
            "type": "temperatureSensor",
            "name": "Temperature",
            "topics": {
                "getCurrentTemperature": "teletask2mqtt/my_teletask/sensor/19/state"
            },
            "history": true,
            "accessory": "mqttthing"
        },`

**Comment by claesmathias** [comment]
I keep on struggling with the new release, rebuilding jeletask container with flag TELETASK_MQTT_RETAINED true and false doesn't get HA to enable the relay lights using auto-discovery. Restarting HA and jeletask containers doesn't help. The dimmable lights work more or less. On the other hand 3.3.6-arm64 works like a charm, not using the RETAINED flag. As you can see in the screenshot below, the relay lights are auto-discovered but cannot be switched on/off.
![Screenshot 2022-09-24 at 14 35 51](https://user-images.githubusercontent.com/1142852/192098631-278475f7-db46-45c3-8286-6d494e87ed4b.png)

**Comment by ridiekel** [comment]
I have released a new version 4.0.3 that contains all the changes. Can you please test taht one?

**Comment by SanderM2** [comment]
Please also check if you have ‘ type: “light” ‘ in your config.json for all relays controlling those lights.

**Comment by claesmathias** [comment]
I don't specify type:"light", only type:"switch". I find type:"light" for sensors in the documentation, I believe that my sockets defined as "switch" are functioning with HA auto-discovery.

Next to this, no possibility to revert back to the original MQTT messages? Looking at spec on https://github.com/arachnetech/homebridge-mqttthing/blob/master/docs/Accessories.md#light-bulb , this is how version 3.3.6-arm64 operated.

**before**
![Screenshot 2022-09-24 at 18 03 20](https://user-images.githubusercontent.com/1142852/192107827-07561525-5470-4c67-9e4b-98bf115fc2eb.png)


**now**
![Screenshot 2022-09-24 at 18 02 01](https://user-images.githubusercontent.com/1142852/192107841-761cb895-adf4-40b7-96e6-bbc3b8887d90.png)

**Comment by SanderM2** [comment]
What happens if you do define type “light” for a relay controlling a light?

**Comment by claesmathias** [comment]
I only set one relay as type "light" and suddenly all of them were enabled in HA. Besides, I guess I will have to adapt the rest my homebridge configuration to use JSON {"state": "100"} for dimmer and temperature sensor?

**Comment by SanderM2** [comment]
I assume all are still not correctly working? Can you turn them all on/off now?

 You might need to  set them to “light” in config.json… i have to dig in the code to find out if this is a bug or not, because the default should be light already.

if it helps by setting it to “light” it’s probably a bug.

**Comment by claesmathias** [comment]
I specified everything as type "light" and adapted homebridge to work with JSON {"state"}, working good for now! How can I easily enable logging in the jeletask container, in case I want to debug teletask FLAG changes?

**Comment by SanderM2** [comment]
I think the reason why you needed to explicitly specify "light" as the type is because ComponentSpech's getType() is null in HARelayConfig.
I added a check for this in PR https://github.com/ridiekel/jeletask/pull/18
Untested but this should fix the "problem" where not specifying the type breaks HA autodiscovery.

**Comment by ridiekel** [comment]
This should be completed in 4.1.0, can you please confirm?

**Comment by ridiekel** [comment]
Fixed since 4.1.0
